### ISSUE TITLE
TC error in existential quantifier about a missing CML context.

### ISSUE THREAD

**Issue by joey-coleman** [symphonytool/symphony#92]
> The following bug was [originally reported](https://sourceforge.net/p/compassresearch/bugs/92/) on Sourceforge by [joeycoleman](https://sourceforge.net/u/joeycoleman/profile/), 2013-10-03 14:31:45.407000:

Initially reported by Jim Woodcock, tested in build based on [f55925b].

The example

```
types
  String2 = seq of char
  Process2 :: id: String2
functions
  findPos(q: seq of Process2,id: String2) pos : nat
    pre  exists p in set (elems q) @ p.id = id
    post true
```

produces this error at "p.id":

  The symbol "(p.id)" requires a CML context to be resolved. No such context is present.

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/92/#ca9f) by [joeycoleman](https://sourceforge.net/u/joeycoleman/profile/), 2013-10-03 14:38:15.283000:

This also fails for expressions like

```
exists p : Process2 @ p.id = id
```

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/92/#b850) by [joeycoleman](https://sourceforge.net/u/joeycoleman/profile/), 2013-10-03 14:39:45.373000:

Probably has something to do with the ASetMultipleBind and ATypeMultipleBind nodes.

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/92/#f794) by [clausbn](https://sourceforge.net/u/clausbn/profile/), 2013-10-07 14:26:26.324000:

There is a problems with binds and expressions. 

This has to do with the way the environment is being shared between the Overture and Compass typechecker. 

No easy fix seems possible. The way the switch between contexts/environments is done, appear to have a fundamental flaw/lack when it comes to binds.

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/92/#a158) by [lausdahl](https://sourceforge.net/u/lausdahl/profile/), 2013-10-08 07:27:50.784000:

The problem here is that overture in the case of "expressions": exists, forall, let. Introduces a new local environment where the definition of the patterns are stored. This is then not handled in CML where the lookup of the pattern is pushed back to CML and where it tried to find a CML env, which doesnt exists. The fix is to first check the overture env before we search for a CML env (that in this case doesnt exit).

It is to be recomented that the very complex double env structure is changed to a simple structure with normal nested env's.

This bug is fixed in: #7e7b6553e3f99807853fc2aabc19e36e49e03084

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/92/#81df) by [lausdahl](https://sourceforge.net/u/lausdahl/profile/), 2013-10-08 07:28:07.724000:
- **status**: open --> closed
